graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,386 @@
+                            return;
+                        }
+                        completed = true;
+                        writeComplete(out, id);
+                        graphqlSubscriptions.remove(id);
+                    }
+                });
+
+        graphqlSubscriptions.put(id, executionResultSubscriber);
+        streamMessage.subscribe(executionResultSubscriber, ctx.eventLoop());
+    }
+
+    private void ensureInitiated() throws Exception {
+        if (!connectionInitiated) {
+            // ConnectionAck not sent yet. Must be closed with 4401 Unauthorized.
+            throw new WebSocketCloseException(4401);
+        }
+    }
+
+    private static String serializeToJson(Object object) throws JsonProcessingException {
+        return mapper.writer().writeValueAsString(object);
+    }
+
+    @Nullable
+    private static String toStringFromJson(@Nullable Object value) throws WebSocketCloseException {
+        if (value == null) {
+            return null;
+        }
+
+        if (value instanceof String) {
+            return (String) value;
+        } else {
+            throw new WebSocketCloseException(4400);
+        }
+    }
+
+    /**
+     * This only works reliably if maybeMap is from Json, as maps(objects) in Json
+     * can only have string keys.
+     */
+    @SuppressWarnings("unchecked")
+    private static Map<String, Object> toMapFromJson(@Nullable Object maybeMap) throws WebSocketCloseException {
+        if (maybeMap == null) {
+            return ImmutableMap.of();
+        }
+
+        if (maybeMap instanceof Map) {
+            final Map<?, ?> map = (Map<?, ?>) maybeMap;
+            if (map.isEmpty()) {
+                return ImmutableMap.of();
+            }
+            return Collections.unmodifiableMap((Map<String, Object>) maybeMap);
+        } else {
+            throw new WebSocketCloseException(4400);
+        }
+    }
+
+    private static <T> T parseJsonString(String content, TypeReference<T> typeReference)
+            throws JsonProcessingException {
+        return mapper.readValue(content, typeReference);
+    }
+
+    private static void writePong(WebSocketWriter out) {
+        out.tryWrite("{\"type\":\"pong\"}");
+    }
+
+    private static void writeConnectionAck(WebSocketWriter out) {
+        out.tryWrite("{\"type\":\"connection_ack\"}");
+    }
+
+    private void writeNext(WebSocketWriter out, String operationId, ExecutionResult executionResult) {
+        final HashMap<String, Object> response = new HashMap<>();
+        response.put("id", operationId);
+        response.put("type", "next");
+        response.put("payload", executionResult.toSpecification());
+        try {
+            final String event = serializeToJson(response);
+            logger.trace("NEXT: {}", event);
+            out.tryWrite(event);
+        } catch (JsonProcessingException e) {
+            logger.error("Error serializing next event", e);
+            writeError(out, operationId, e);
+        }
```

Comment:
It seems like we shouldn't catch the exception here but propagate to the caller so that `subscription.cancel()` is called.
Or, how about renaming this to `tryWriteNext` and returning false when it fails? If so the caller can cancel the subscription.

Reply:
Changed to propagate the exception to the caller.